# CoVarFinder
---This title is under consideration! It's super basic but at the moment I can't think of anything elseCancel changes

This code is in a very primitive state but it's still very usefull for those who know python basics. All you need is to preproccess your .bam file with the snpEff applications in java. Then the script reads all the genomic mutations, translates them in aminoacid alterations and compares them with all the defining mutations of Variants of Concern. The final output is a list of all uniwue mutation, but you can call the shared mutation as well. In the future I want to expand to other variants as well and to variant frequencies which will be a bitch but I will figure!
